feat: Add fontFamily and fontSrc

jadutter opened the pull request:
Add fontFamily and fontSrc option to text layout.

Allow merchants to pass fontFamily and fontSrc options in order to customize the font for upstream messages.

jadutter committed:
Add fontFamily and fontSrc to server/ for text layouts

jadutter committed:
Change fontFamily and fontSrc to work on other layouts. textSize set for text layout only at the moment.

jadutter committed:
Changed fontSrc to accept string or array of strings, as well as provide font format in the css rule

jadutter committed:
Switched to validate fontSrc urls with regexp

jadutter committed:
Refined regex

jadutter committed:
Added validation for fontSrc before sending query to cpnw

merlinstardust commented on server/locale/DE/validOptions.js:
```diff
@@ -8,13 +8,20 @@ export default {
+            // size: [Types.NUMBER, [12, 10, 11, 13, 14, 15, 16]],
+            fontFamily: [Types.STRING],
+            fontSrc: [Types.MULTI_STRING]
+        }
```
For flex, I think we can simplify this to just have it be `flex.fontFamily` and `flex.fontSrc`. Color options can just be handled at `flex.color`.

merlinstardust replied:
`fontSrc` should be fully spelled out as `fontSource`. It's better to use full spellings unless the abbreviation is extremely long.

merlinstardust commented on server/locale/GB/validOptions.js:
```diff
@@ -9,13 +9,20 @@ export default {
+            color: [Types.STRING, ['black', 'white', 'monochrome', 'grayscale|greyscale']],
+            // size: [Types.NUMBER, [12, 10, 11, 13, 14, 15, 16]],
+            fontFamily: [Types.STRING],
+            fontSrc: [Types.MULTI_STRING]
```
As above, we can flatten this and not need `flex.text`, which could also be confusing

jadutter replied:
I’ll admit, I found it confusing too. It took me a bit to recognize we were structuring the object default slightly differently than the input object we’re comparing it against. I figured it was better to follow the convention we were using.

If we moved flex.text.fontSource to flex.fontSource, then I think text.text.fontSource should likewise move to text.fontSource.

Ditto for fontFamily.

merlinstardust commented on server/locale/US/GPL/validOptions.js:
```diff
@@ -19,11 +20,23 @@ export default {
+            color: [Types.STRING, ['black', 'white', 'monochrome', 'grayscale|greyscale']],
+            // size: [Types.NUMBER, [12, 10, 11, 13, 14, 15, 16]],
+            fontFamily: [Types.STRING],
+            fontSrc: [Types.MULTI_STRING]
```
As above, we can flatten this and not need `flex.text`

merlinstardust commented on server/locale/US/GPL/validOptions.js:
```diff
@@ -19,11 +20,23 @@ export default {
+            color: [Types.STRING, ['black', 'white', 'monochrome', 'grayscale|greyscale']],
+            size: [Types.NUMBER, [12, 10, 11, 13, 14, 15, 16]],
+            fontFamily: [Types.STRING],
+            fontSrc: [Types.MULTI_STRING]
```
We shouldn't need these fields in custom, since custom gets defined with a URL pointing to custom HTML files. Ideally, if they're using `fontFamily` or `fontSrc`, they won't need to repeat it in custom

merlinstardust commented on server/locale/US/PPC/validOptions.js:
```diff
@@ -19,11 +20,23 @@ export default {
+            color: [Types.STRING, ['black', 'white', 'monochrome', 'grayscale|greyscale']],
+            // size: [Types.NUMBER, [12, 10, 11, 13, 14, 15, 16]],
+            fontFamily: [Types.STRING],
+            fontSrc: [Types.MULTI_STRING]
```
As above, we can flatten this and not need `flex.text`

merlinstardust commented on server/locale/US/PPC/validOptions.js:
```diff
@@ -19,11 +20,23 @@ export default {
+            color: [Types.STRING, ['black', 'white', 'monochrome', 'grayscale|greyscale']],
+            size: [Types.NUMBER, [12, 10, 11, 13, 14, 15, 16]],
+            fontFamily: [Types.STRING],
+            fontSrc: [Types.MULTI_STRING]
```
As above, custom shouldn't need these

merlinstardust commented on server/message/index.jsx:
```diff
@@ -36,6 +36,34 @@ const applyCascade = curry((style, flattened, type, rules) =>
         type === Array ? [] : {}
     )
 );
+const getFontSrcRule = fontSrc => {
```
This function is getting repeated a lot here and in `server/validateStyle` and in `src/zoid/message/validation`, so I think it should be abstracted to not repeat as much code.

merlinstardust replied:
Add newline before the function

merlinstardust commented on server/message/index.jsx:
```diff
@@ -36,6 +36,34 @@ const applyCascade = curry((style, flattened, type, rules) =>
+            .map(url => {
+                const match = urlPat.exec(url);
+                const ext = match ? match[2] : null;
+                const fmt = fontFormats[ext];
```
`ext` and `fmt` should be `extension` and `format`

merlinstardust commented on server/message/index.jsx:
```diff
@@ -36,6 +36,34 @@ const applyCascade = curry((style, flattened, type, rules) =>
+        ttf: 'ttf', // truetype
+        otf: 'otf' // opentype
+        // eot: 'eot', // embedded opentype
+        // svg: 'svg' // svg
```
Since we're ignoring `woff2`, `eot`, and `svg`, these can be removed.

merlinstardust commented on server/message/index.jsx:
```diff
@@ -36,6 +36,34 @@ const applyCascade = curry((style, flattened, type, rules) =>
+        // eot: 'eot', // embedded opentype
+        // svg: 'svg' // svg
+    };
+    if (fontSrc?.constructor === Array) {
```
This check shouldn't be necessary since we're already forcing an array at the validate step

merlinstardust commented on server/message/index.jsx:
```diff
@@ -62,14 +90,40 @@ export default ({ options, markup, locale }) => {
+    }
+    if (fontFamilyRule || textSizeRule) {
+        customFontStyleRules.push(`${fontSelector}{ ${fontFamilyRule}${textSizeRule} }`);
+    }
```
All these font lines have made it more difficult to read. I think moving all of it in to a function will make it more readable

merlinstardust commented on server/message/index.jsx:
```diff
@@ -62,14 +90,40 @@ export default ({ options, markup, locale }) => {
+    const fontSelector = `
+.message__messaging, 
+.message__messaging .message__headline span, 
+.message__messaging .message__disclaimer span`;
```
Indent these and move the closing backtick to the next line so it looks cleaner

merlinstardust commented on server/message/index.jsx:
```diff
@@ -62,14 +90,40 @@ export default ({ options, markup, locale }) => {
+            @font-face {
+                font-family: '${fontFamily}'; 
+                ${fontSrcRule}
+            }`);
```
This can fit on one line

merlinstardust commented on server/validateStyle.js:
```diff
@@ -14,7 +14,38 @@ const logInvalidOption = (addLog, location, options, val) =>
         }".`
     );
-
+function validateFontSrc(fontSrc) {
```
Add newline before and after the function

merlinstardust commented on server/message/index.jsx:
```diff
@@ -62,14 +90,40 @@ export default ({ options, markup, locale }) => {
+    const textSizeRule = textSize ? `font-size: ${textSize}px; ` : '';
+    const fontFamilyRule = fontFamily
+        ? `font-family: '${fontFamily}', PayPal-Sans-Big, PayPal-Sans, Arial, sans-serif; `
+        : '';
```
In the case that a merchant wants to use a generic font family, then `fontFamily` should not have quotes as generic families are keywords. It should still have quotes when not generic.

Here's an array of generic font-families I found on MDN: https://developer.mozilla.org/en-US/docs/Web/CSS/font-family

```
['serif', 'sans-serif', 'monospace', 'cursive', 'fantasy', 'system-ui', 'ui-serif', 'ui-sans-serif', 'ui-monospace', 'math', 'emoji', 'fangsong']
```

jadutter replied:
I'll add a function to check if they're using one of those generic names.

merlinstardust commented on src/zoid/message/validation.js:
```diff
@@ -103,6 +103,33 @@ export default {
+                });
+            }
+            if (fontSrc) {
+                console.log(`fontSrc:`, fontSrc);
```
Remove console.log artifacts

merlinstardust commented on server/validateStyle.js:
```diff
@@ -42,7 +73,18 @@ function getValidVal(addLog, typeArr, val, location) {
+                return [val];
+            }
+            if (val?.constructor === Array) {
+                console.log(val);
```
Remove console.log artifacts

merlinstardust reviewed (changes requested):
The code itself looks pretty good. I'm still having trouble trying to get a fontSrc to load on my machine. Would be worth having another dev or two also try to test it on their machine.

Also, be sure to commit via the command line, so the pre-commit linting check runs. I found the following after running `npm run lint`

```
server/validateStyle.js
  29:24  error    'fontSrc' is already declared in the upper scope  no-shadow
  84:17  warning  Unexpected console statement                      no-console

src/zoid/message/validation.js
  127:17  warning  Unexpected console statement                          no-console
  128:17  error    Assignment to property of function parameter 'style'  no-param-reassign
  130:24  error    Assignment to property of function parameter 'style'  no-param-reassign

✖ 5 problems (3 errors, 2 warnings)
```

jadutter committed:
simplified fontSrc by changing it to stringify its value & encode as base64, then decode and validate it is a string for a url server side.

jadutter committed:
limited fontSrc to 10 urls.

jadutter committed:
Renamed fontSrc to fontSource.
Changed fontFamily to accept generic family names.
Refactored to not need to disable no-param-reassign.
Removed superfluous options from flex and custom (in validOptions files).

jadutter committed:
fixed some bugs pertaining to renaming fontSrc, and refactoring validation.js

jadutter committed:
Added logging for fontFamily and fontSource bsaed on validateStyle. Added regexp to help confirm fontFamily is a valid value

jadutter committed:
Changed fontSource to no longer use base64

jadutter committed:
Added some tests for SSR style options

jadutter committed:
Modified fontSource to accept all font types. It will be up to merchants to provide extra urls for fallbacks.

jadutter committed:
Fixed the regex to use the declared font formats

jadutter committed:
Fixed it so fontFamily and fontSource can be given as inline html, and moved font css rules into their own function for organization

Seavenly commented on server/message/index.jsx:
```diff
@@ -36,8 +40,111 @@ const applyCascade = curry((style, flattened, type, rules) =>
+        error: null
+    };
+    try {
+        const fontSource = val;
```
Can the parameter `val` just be renamed to `fontSource`?

jadutter replied:
Yes. In previous commits, I was changing what was stored in fontSource, so I was trying not to modify the input parameter. I'll simplify this.

Seavenly commented on server/message/index.jsx:
```diff
@@ -36,8 +40,111 @@ const applyCascade = curry((style, flattened, type, rules) =>
+        eot: 'eot', // embedded opentype
+        svg: 'svg' // svg
+    };
+    const urlPattern = RegExp(`^(?<url>https?://.+?.(?<ext>${[...Object.values(fontFormats)].join('|')}))$`);
```
It doesn't look like we're actually using the named capture groups. Would it be sufficient without them? Also, are we not supporting relative paths?

jadutter replied:
No, we wouldn't want to support relative pathing. Since this goes into our iframe, it would grab the path relative to our iframe. 
To support relative urls, we'd likely need to modify it so that fontSource values get checked and appended with the page origin before the request went out to cpnw. 

I'd initially tried pulling out the named capture groups, but it seems they're not supported in regular babel. I wasn't sure if it was acceptable to add `@babel/plugin-transform-named-capturing-groups-regex` to the sdk, so I refactored it to pull out the items by index. 
I'd be more than happy to use named capture groups for this instead.

jadutter replied:
Keeping this as is since we cannot add the babel plugin for regex named groups.

Seavenly commented on server/message/index.jsx:
```diff
@@ -36,8 +40,111 @@ const applyCascade = curry((style, flattened, type, rules) =>
+    try {
+        const fontSource = val;
+        payload.fontSource = fontSource;
+        const ruleVal = (fontSource?.constructor === Array ? fontSource : [fontSource])
```
Consider using the more declarative `Array.isArray()`

jadutter replied:
I made the change to my local copy after you mentioned the same thing on cpnw.

Seavenly commented on server/message/index.jsx:
```diff
@@ -36,8 +40,111 @@ const applyCascade = curry((style, flattened, type, rules) =>
+        payload.fontSource = fontSource;
+        const ruleVal = (fontSource?.constructor === Array ? fontSource : [fontSource])
+            .slice(0, 10)
+            .filter(e => typeof e === 'string')
```
At this point, these input aspects should have been vetted. Can these be removed?

Seavenly commented on server/message/index.jsx:
```diff
@@ -36,8 +40,111 @@ const applyCascade = curry((style, flattened, type, rules) =>
+            .slice(0, 10)
+            .filter(e => typeof e === 'string')
+            .map(url => {
+                const match = urlPattern.exec(url);
+                const extension = match ? match[2] : null;
```
```suggestion
                const extension = urlPattern.exec(url)?.[2];
```

jadutter replied:
Updated.

Seavenly commented on server/message/index.jsx:
```diff
@@ -36,8 +40,111 @@ const applyCascade = curry((style, flattened, type, rules) =>
+                const match = urlPattern.exec(url);
+                const extension = match ? match[2] : null;
+                const format = fontFormats[extension];
+                if (url && format) {
```
We should be able to safely assume `url` exists at this point.
``` suggestion
                if (format) {
```

Seavenly commented on server/message/index.jsx:
```diff
@@ -36,8 +40,111 @@ const applyCascade = curry((style, flattened, type, rules) =>
+                }
+                return '';
+            })
+            .filter(e => e)
```
```suggestion
            .filter(Boolean)
```

jadutter replied:
Updating here, and other locations where I was filtering for truthy values.

Seavenly commented on server/message/index.jsx:
```diff
@@ -36,8 +40,111 @@ const applyCascade = curry((style, flattened, type, rules) =>
+    const payload = {
+        fontSource: val,
+        validValues: [],
+        error: null
```
Can we just omit the `error` value completely unless it exists?

Seavenly commented on server/message/index.jsx:
```diff
@@ -36,8 +40,111 @@ const applyCascade = curry((style, flattened, type, rules) =>
+        validValue: null
+    };
+    /* eslint-disable-next-line security/detect-unsafe-regex */
+    const fontPattern = RegExp(/^(?<font>([a-zA-Z0-9._-]\s*)+?)$/);
```
Do we need to validate the font here when it should have been vetted prior to this?

jadutter replied:
We don't need it anymore. I'll remove it.

Seavenly commented on server/message/index.jsx:
```diff
@@ -36,8 +40,111 @@ const applyCascade = curry((style, flattened, type, rules) =>
+    };
+    /* eslint-disable-next-line security/detect-unsafe-regex */
+    const fontPattern = RegExp(/^(?<font>([a-zA-Z0-9._-]\s*)+?)$/);
+    const genericFamilies = {
+        serif: 'serif',
+        'sans-serif': 'sans-serif',
+        monospace: 'monospace',
+        cursive: 'cursive',
+        fantasy: 'fantasy',
+        'system-ui': 'system-ui',
+        'ui-serif': 'ui-serif',
+        'ui-sans-serif': 'ui-sans-serif',
+        'ui-monospace': 'ui-monospace',
+        math: 'math',
+        emoji: 'emoji',
+        fangsong: 'fangsong'
+    };
```
It looks like this is needed because the generic font families can't be surrounded with `'`s, but specific font names should be to capture spaces. Can we add a comment explaining this reasoning Do we need to have `math`, `emoji`, and `fangsong` in the list?

jadutter replied:
This list was pulled from MDN's list of generic names, without further thought. 

I think we can safely drop `math` and `emoji`, but we may want to retain `fangsong` to future proof needing to support Chinese fonts.

Seavenly commented on server/message/index.jsx:
```diff
@@ -36,8 +40,111 @@ const applyCascade = curry((style, flattened, type, rules) =>
+        emoji: 'emoji',
+        fangsong: 'fangsong'
+    };
+    const fontFamily = typeof val === 'string' ? genericFamilies[val] ?? `'${val}'` : '';
```
It should be safe to assume `val` is a string from prior vetting.

Seavenly commented on src/utils/elements.js:
```diff
@@ -37,15 +37,31 @@ export function isElement(el) {
+    };
+    const getOptionValue = (name, value) => {
+        if (name === 'style-text-fontSource' && value?.indexOf('[') >= 0) {
+            try {
+                return flattenedToObject(name, JSON.parse(value));
+            } catch (err) {
+                console.error(err.stack);
+            }
+            try {
+                return flattenedToObject(name, JSON.parse(value.replace(/'/g, '"')));
+            } catch (err) {
+                console.error(err.stack);
+            }
```
Instead of 2 try blocks consider just using `return flattenedToObject(name, JSON.parse(value.replace(/'/g, '"')));` since it will cover both cases.

Seavenly commented on src/utils/elements.js:
```diff
@@ -37,15 +37,31 @@ export function isElement(el) {
+        'data-pp-style-text-fontsource': 'data-pp-style-text-fontSource'
+    };
+    const getOptionValue = (name, value) => {
+        if (name === 'style-text-fontSource' && value?.indexOf('[') >= 0) {
```
Consider using `String.prototype.startsWith` as opposed to `indexOf() >= 0`. It requires importing the polyfill manually from `core-js-pure` but it's generally preferred due to its declarative nature.

jadutter committed:
Generic optimizations and code clean up

jadutter committed:
Comments and optimization revisions

jadutter committed:
Removed a console.error

Seavenly commented:
:tada: This PR is included in version 1.17.0 :tada:

The release is available on [GitHub release](https://github.com/paypal/paypal-messaging-components/releases/tag/v1.17.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
